app/models/manageiq/providers/amazon/cloud_manager/event_target_parser.rb

```diff
@@ -0,0 +1,122 @@
+                      :floating_ips
+                    end
+
+    add_target!(target_collection, target_class, resource_id) if target_class && resource_id
+  end
+
+  def collect_cloudwatch_api_references!(target_collection, event_data, depth = 0)
+    # Check a reasonable depth, so this can't fail with max stack size
+    raise "Depth 20 reached when scanning EmsEvent for Targets" if depth > 20
+
+    # Cloud
+    add_target!(target_collection, :vms, event_data["instanceId"]) if event_data["instanceId"]
+    add_target!(target_collection, :miq_templates, event_data["imageId"]) if event_data["imageId"]
+    add_target!(target_collection, :key_pairs, event_data["keyName"]) if event_data["keyName"]
+    add_target!(target_collection, :orchestration_stacks, event_data["stackId"]) if event_data["stackId"]
+    add_target!(target_collection, :orchestration_stacks, event_data["stackName"]) if event_data["stackName"]
+    # Network
+    add_target!(target_collection, :cloud_networks, event_data["vpcId"]) if event_data["vpcId"]
+    add_target!(target_collection, :cloud_subnets, event_data["subnetId"]) if event_data["subnetId"]
+    add_target!(target_collection, :network_ports, event_data["networkInterfaceId"]) if event_data["networkInterfaceId"]
+    add_target!(target_collection, :security_groups, event_data["groupId"]) if event_data["groupId"]
+    add_target!(target_collection, :floating_ips, event_data["allocationId"]) if event_data["allocationId"]
+    add_target!(target_collection, :load_balancers, event_data["loadBalancerName"]) if event_data["loadBalancerName"]
+    # Block Storage
+    add_target!(target_collection, :cloud_volumes, event_data["volumeId"]) if event_data["volumeId"]
+    add_target!(target_collection, :cloud_volume_snapshots, event_data["snapshotId"]) if event_data["snapshotId"]
+
+    # TODO(lsmola) how to handle tagging? Tagging affects e.g. a name of any resource, but contains only a generic
+    # resourceID
+    # "requestParameters"=>
+    #   {"resourcesSet"=>{"items"=>[{"resourceId"=>"vol-07ad036724e3175a5"}]},
+    #    "tagSet"=>{"items"=>[{"key"=>"Name", "value"=>"ladas_volue_2"}]}},
+    # I think we can parse the resource id, so guess where it belongs.
+    # TODO(lsmola) RegisterImage, by creating image from a volume snapshot, should we track the block device
+    # mapping and refresh also snapshot?
+
+    # Collect nested references
+    collect_cloudwatch_api_references!(target_collection, event_data["networkInterface"], depth + 1) if event_data["networkInterface"]
+
+    (event_data.fetch_path("groupSet", "items") || []).each do |x|
+      collect_cloudwatch_api_references!(target_collection, x, depth + 1)
+    end
+    (event_data.fetch_path("instancesSet", "items") || []).each do |x|
+      collect_cloudwatch_api_references!(target_collection, x, depth + 1)
+    end
+    (event_data.fetch_path("instances") || []).each do |x|
+      collect_cloudwatch_api_references!(target_collection, x, depth + 1)
```

Comment:
would it make sense to expand this api? `collect_cloud_watch_api_referenceses!`

of course, naming is hard. Then again, do you ever `collect_cloudwatch_api_references` with a single array?
```
collect_cloud_watch_api_referenceses!(target_collection, event_data.fetch_path("instances") || [], depth + 1)
```